engine/lib/phx/src/audio/sound.rs

```diff
@@ -1,500 +1,164 @@
+        EndPosition, PlaybackPosition, Region,
+    },
+    spatial::emitter::EmitterHandle,
+    tween::{Easing, Tween},
+    StartTime,
+};
 
-use fmod_sys::*;
-
-#[derive(Copy, Clone)]
-#[repr(C)]
 pub struct Sound {
-    pub desc: *mut SoundDesc,
-    pub handle: *mut FMOD_CHANNEL,
-    pub state: SoundState,
-    pub autoPos: *const Vec3,
-    pub autoVel: *const Vec3,
-    pub freeOnFinish: bool,
+    path: String,
+    sound_data: StaticSoundData,
+    emitter: Option<EmitterHandle>,
+    sound_handle: Option<StaticSoundHandle>,
 }
-pub type SoundState = u8;
 
-#[inline]
-pub fn FMODCALL(result: FMOD_RESULT) {
-    if result != FMOD_RESULT::FMOD_OK {
-        panic!(
-            "FMOD operation failed with {:?} ({})",
-            result,
-            result.error_string()
-        );
+impl Sound {
+    pub fn has_sound_handle(&self) -> bool {
+        self.sound_handle.is_some()
     }
-}
 
-unsafe extern "C" fn Sound_Callback(
-    channel: *mut FMOD_CHANNELCONTROL,
-    _controlType: FMOD_CHANNELCONTROL_TYPE,
-    callbackType: FMOD_CHANNELCONTROL_CALLBACK_TYPE,
-    _a: *mut libc::c_void,
-    _b: *mut libc::c_void,
-) -> FMOD_RESULT {
-    if callbackType == FMOD_CHANNELCONTROL_CALLBACK_TYPE::FMOD_CHANNELCONTROL_CALLBACK_END {
-        let mut this: *mut Sound = std::ptr::null_mut();
-        FMODCALL(FMOD_Channel_GetUserData(
-            channel as *mut FMOD_CHANNEL,
-            &mut this as *mut *mut Sound as _,
-        ));
-        Sound_SetState(&mut *this, 4 as SoundState);
+    pub fn sound_data(&self) -> &StaticSoundData {
+        &self.sound_data
     }
-    FMOD_RESULT::FMOD_OK
-}
 
-#[inline]
-unsafe extern "C" fn Sound_EnsureLoadedImpl(this: &mut Sound, func: *const libc::c_char) {
-    if this.state as i32 == 1 {
-        SoundDesc_FinishLoad(&mut *this.desc, func);
-        FMODCALL(FMOD_System_PlaySound(
-            Audio_GetHandle() as *mut FMOD_SYSTEM,
-            (*this.desc).handle,
-            std::ptr::null_mut(),
-            1,
-            &mut this.handle,
-        ));
-        FMODCALL(FMOD_Channel_SetUserData(
-            this.handle,
-            this as *mut Sound as *mut _,
-        ));
-        FMODCALL(FMOD_Channel_SetCallback(
-            this.handle,
-            Some(
-                Sound_Callback
-                    as unsafe extern "C" fn(
-                        *mut FMOD_CHANNELCONTROL,
-                        FMOD_CHANNELCONTROL_TYPE,
-                        FMOD_CHANNELCONTROL_CALLBACK_TYPE,
-                        *mut libc::c_void,
-                        *mut libc::c_void,
-                    ) -> FMOD_RESULT,
-            ),
-        ));
-        Sound_SetState(this, 2 as SoundState);
-        if Sound_Get3D(this) {
-            let mut zero: Vec3 = Vec3::ZERO;
-            Sound_Set3DPos(this, &mut zero, &mut zero);
-        }
+    pub fn set_emitter(&mut self, emitter: EmitterHandle) {
+        // FIXME: for some reason adding emitter kills the sound
```

Comment:
Does this mean this function doesnt work?

Reply:
Yes, setting emitter doesn't work but without it sound plays ok, so I'll postpone investigation of this issue for later.